### DISCUSSION TITLE
Make a Loop Option for Song Arranger Mode

### DISCUSSION THREAD

**Discussion by Zappa711** [SynthstromAudible/DelugeFirmware#301] [Ideas]
Please include an option, like press shift play or something while in Arranger mode, to start playback of the arrangement which will loop itself when reaching the end and start seamlessly from the beginning. This will make big steps forward in using this mode for live performances.

**Comment by PaulFreund** [comment]
Wonderful idea, I have this on my longlist! I moved the idea to the Discussions tab in accordance with our [Contributors guidelines](https://github.com/SynthstromAudible/DelugeFirmware/blob/community/CONTRIBUTING.md) :)

**Comment by weavermedia** [comment]
Yes, great idea! On my list too. I'd like to see options for:
- Loop the whole arrangement (as mentioned)
- Loop the visible area shown on the grid pads, and respond when the displayed area changes
- Set custom loop start and loop end points to loop a specific section

**Reply by sichtbeton** [reply]
Yes please. All of the above

**Reply by WillStefan** [reply]
Seconded - At the bare minimum custom loop start and end points, and ideally the ability to support multiple sections.  Maybe even something like a custom "section" track that could be used to configure and select the active sections.

**Reply by lunerfox** [reply]
I'd definitely use this; I'm often playing other instruments and it would be very nice to step through a progression without having very long clips. This is especially useful for arranging audio, since it doesn't have iteration dependence to keep things interesting

**Comment by manysounds** [comment]
Yes, having an arrangement with a loopable section or sections for remix/jam/etc ala Ableton would be amazing

**Comment by manysounds** [comment]
100% remix tool yes!

**Comment by sichtbeton** [comment]
Adding to that: copy and paste the current loop. Would make for a great addition to the arranger.

**Comment by duncanrmi** [comment]
you can already drop out of the arranger view back into song mode & have a section loop around, but being able to set it up in the arranger would be fab, yes.
take a look at the way the arranger works in the octatrack- same sort of thing, I think. 
both song & arranger views are lacking functionality & visual feedback atm, & I also like the idea of a 'master' track to control various attributes of the song playback. 
this could contain 'macro' functionality such as the switching in & out of clips, the behaviour of the playback sequencing (multi-bar looping from x1 to 'infinite until interrupted somehow'), but also stuff like mix & parameter automation on a clip &/or 'affects all' basis.
I haven't bottomed out all of the details in my imagination, but I'm thinking along the lines of the master track in the emu sequencers (such as in the XL7) or the yamaha qy series walkstations.
so the top row of the grid (any length you like, but the sixteen in the current zoom view) would be centred on the stage currently playing.
the other seven rows could contain macro & mix/parameter info for the stage- number of iterations, 'affects all' & clip-level parameters.
shift-tapping any of these would mute the effect (pad goes dark) so that the values are held at whatever they are when the pad is muted.

how I imagine it: each pad on the grid in arranger view would be a stage in the arrangement, & _tapping_ the pad would drop you into 'song' view for that stage. 
this stage would be forced to loop while you are 'in' it, editing it or whatever. maybe this behaviour could be configurable, & do something clever with 'cross-screen' so that you can either stay 'in' the stage you were editing or have it switch to the currently playing stage.

any edits you make that correspond to aspects of the stage that are under 'arranger' control would update the corresponding arranger control pad.
_holding_ the pad in arranger view would let you set the number of iterations of that entire stage, from x1 to 'hold'.
you'd advance to the next stage by _pressing the current pad & tapping the next pad that you want to play_. 
holding shift at the same time would do the immediate jump instead of waiting until the loop point.
left to its own devices, the deluge would just work its way through the stages in order, or sit on the first one where the iterations are set to  'hold'. 

I sense I'm beginning to ramble.... :-)

**Comment by lobit12** [comment]
Ok so I've been on this train..  100% what Ive been thinking about.  like a performance orientated version of playing your 'song' back but being able to sort of live-remix in the process.

What I found is this is already possible to a degree: 

  Im playing my song back and while currently can't do the '<> press + play' command to bounce around (if you try to do it you get the message 'CANT' right now) , BUT If I go into clips view, play a few clips,  which will pop the deluge into clips play mode, essentially taking over from song playback,  then if I go back into song and find the bar number/ spot I want to drop into and then do '<>press +play' command , it will drop seamlessly into that bar on the downbeat = fantastic. okay halfway there.  

Now, what would be great is if it I could, first off, when I go into clips mode from song, get a visual indication of what clips are playing at that moment in that part of the song, (currently when you do that the clips are showing all off, which I guess in a way they are at that moment since its kindof between modes at this point)  If I knew which clips were active somehow, I could trigger the correct ones at that moment so that would avoid that noticeable transition from one mode to the other. 

And obviously second would be if you could just skip around within song mode without having to go back and forth if you didn't want to. 

But that said I could see alot of time the flow would go something like this: you are playing a section of the song as it was 'recorded' then you want to add or subtract some clips in grid view then pop back in at say another section of the song, so you can kindof already do that but when you go from one mode to the next you have to kind of know what clips are playing in the background because theres currently no indication of that. I tend to have a rough idea but since its just coloured blocks sometimes I forget in the moment.

I can make a little vid of how Im currently making this work if anyone is curious.

Being able to loop a chunk easily would of course be amazing, I had pictured sortof the way it looks in the sample editor where you would have different colored lines for loop start/ end point and you could dynamically move the 'window' between them quantized to the bar.

**Comment by wargoth** [comment]
I find it very productive to draft music ideas on Deluge. Loops in arrangement view would help to work on transitions. 

It would be great to have the top row in arrangement view (maybe always visible regardless of scroll position) dedicated to loop handles which could be set up just like any other clip (maybe with a default length of 8 bars, and/or memorizing previously set length ). It could appear in rainbow colors to distinguish it from other clips. 

Loops in arrangement view could open other applications like:

1. Resampling (SHIFT+RECORD) could also automatically stop at loop ends making it easier to get quantized recordings
2. A function to instantly convert loop into a song section (RECORD+SONG shortcut). Great for cutting up stems into song loops
3. I often use arrangement view for sampling external MIDI instruments. This could enable quantized audio recordings and overdubs.

**Reply by WillStefan** [reply]
Yes, this is precisely what I'm hoping for - having the ability to define and select selects loop points would be a game changer.

**Comment by wargoth** [comment]
Any chance this will be picked up for the upcoming Chopin release?

**Reply by seangoodvibes** [reply]
> Any chance this will be picked up for the upcoming Chopin release?

Chopin release features are closed. If someone works on this it will be part of the nightly build.
